StitchCore/StitchCoreTests/Admin/Resources.swift

```diff
@@ -0,0 +1,193 @@
+        return self.httpClient.doRequest {
+            $0.endpoint = self.url
+            }.flatMap {
+                return try JSONDecoder().decode([Model].self,
+                                                from: JSONSerialization.data(withJSONObject: $0))
+        }
+    }
+}
+
+/// Adds an endpoint method that GETs some id
+internal protocol Gettable: Resource { associatedtype Model: Decodable }
+extension Gettable {
+    public func get() -> Promise<Model> {
+        return self.httpClient.doRequest {
+            $0.endpoint = self.url
+            }.flatMap {
+                return try JSONDecoder().decode(Model.self,
+                                                from: JSONSerialization.data(withJSONObject: $0))
+        }
+    }
+}
+
+/// Adds an endpoint method that DELETEs some id
+internal protocol Removable: Resource {}
+extension Removable {
+    public func remove() -> Promise<Any> {
+        return self.httpClient.doRequest {
+            $0.endpoint = self.url
+            $0.method = .delete
+        }
+    }
+}
+
+/// Adds an endpoint method that POSTs new data
+internal protocol Creatable: Resource {
+    associatedtype CreatorModel: Encodable
+    associatedtype Model: Decodable
+}
+extension Creatable {
+    public func create(data: CreatorModel) -> Promise<Model> {
+        return self.httpClient.doRequest {
+            $0.endpoint = self.url
+            $0.method = .post
+            try $0.encode(withData: data)
+            }.flatMap {
+                return try JSONDecoder().decode(Model.self,
+                                                from: JSONSerialization.data(withJSONObject: $0))
+        }
+    }
+}
+
+/// Adds an endpoint method that PUTs some data
+internal protocol Updatable: Resource { associatedtype Model: Codable }
+extension Updatable {
+    public func update(data: Model) -> Promise<Model> {
+        return self.httpClient.doRequest {
+            $0.endpoint = self.url
+            $0.method = .put
+            try $0.encode(withData: data)
+            }.flatMap {
+                return try JSONDecoder().decode(Model.self,
+                                                from: JSONSerialization.data(withJSONObject: $0))
+        }
+    }
+}
+
+/// Adds an endpoint that enables a given resource
+internal protocol Enablable: Resource {}
+extension Enablable {
+    public func enable() -> Promise<Any> {
+        return self.httpClient.doRequest {
+            $0.endpoint = "\(self.url)/enable"
+            $0.method = .put
+        }
+    }
+}
+
+/// Adds an endpoint that disables a given resource
+internal protocol Disablable: Resource {}
+extension Disablable {
+    public func disable() -> Promise<Any> {
+        return self.httpClient.doRequest {
+            $0.endpoint = "\(self.url)/disable"
+            $0.method = .put
+        }
+    }
+}
+
+/// Resource that lists the current groupId's applications
+internal final class AppsResource: BasicResource, Listable {
```

Comment:
[style q] This structure is nice and a lot more understandable, but for the sake of making it cleaner should we just drop the "Resource" suffix so referring to the Rule class here is `Apps.App.Services.Rules.Rule`? The suffix seems redundant now that everything inherits `BasicResource and it might make things elsewhere look cleaner.